src/tasks/TaskHelper.ts

```diff
@@ -0,0 +1,166 @@
+
+    if (allTasks.some(t => t.label === task.label)) {
+        return false;
+    }
+
+    allTasks.push(task);
+    await workspaceTasks.update('tasks', allTasks);
+    return true;
+}
+
+export async function getAssociatedDockerRunTask(debugConfiguration: DockerDebugConfiguration): Promise<DockerRunTaskDefinition | undefined> {
+    // Using config API instead of tasks API means no wasted perf on re-resolving the tasks, and avoids confusion on resolved type !== true type
+    const workspaceTasks = workspace.getConfiguration('tasks');
+    const allTasks: TaskDefinition[] = workspaceTasks && workspaceTasks.tasks as TaskDefinition[] || [];
+
+    return await recursiveFindTaskByType(allTasks, 'docker-run', debugConfiguration);
+}
+
+export async function getAssociatedDockerBuildTask(runTask: DockerRunTaskDefinition): Promise<DockerBuildTaskDefinition | undefined> {
+    // Using config API instead of tasks API means no wasted perf on re-resolving the tasks, and avoids confusion on resolved type !== true type
+    const workspaceTasks = workspace.getConfiguration('tasks');
+    const allTasks: TaskDefinition[] = workspaceTasks && workspaceTasks.tasks as TaskDefinition[] || [];
+
+    return await recursiveFindTaskByType(allTasks, 'docker-build', runTask);
+}
+
+export async function getOfficialBuildTaskForDockerfile(dockerfile: string, folder: WorkspaceFolder): Promise<Task | undefined> {
+    let buildTasks = await tasks.fetchTasks({ type: 'docker-build' });
+    buildTasks =
+        buildTasks.filter(t => t.execution.args.some(a => { // Find all build tasks where an argument to 'docker build' is this Dockerfile
+            let arg: string;
+            if (typeof a === 'string') {
+                arg = a;
+            } else {
+                arg = a.value;
+            }
+
+            arg = resolveWorkspaceFolderPath(folder, arg);
+            return arg === dockerfile;
+        }));
+
+    if (buildTasks.length === 1) {
+        return buildTasks[0]; // If there's only one build task, take it
+    } else if (buildTasks.length > 1) {
+        return buildTasks.find(t => t.name === 'docker-build: release') || buildTasks[0]; // If there's multiple try finding one with the name 'docker-build: release', else take first
+    }
+
+    return undefined;
+}
+
+export function resolveWorkspaceFolderPath(folder: WorkspaceFolder, folderPath: string): string {
+    return folderPath.replace(/\$\{workspaceFolder\}/gi, folder.uri.fsPath);
+}
+
+export function unresolveWorkspaceFolderPath(folder: WorkspaceFolder, folderPath: string): string {
+    // tslint:disable-next-line: no-invalid-template-strings
+    return folderPath.replace(folder.uri.fsPath, '${workspaceFolder}').replace(/\\/g, '/');
+}
+
+// tslint:disable-next-line: no-any
+async function recursiveFindTaskByType(allTasks: TaskDefinition[], type: string, node: any): Promise<TaskDefinition | undefined> {

```

Comment:
>recursiveFindTaskByType [](start = 15, length = 23)

This logic is very difficult to understand.  Why structure it in this way?

Reply:
You need to recursively follow down your dependency chain in order to find the task you want. The dependency chain can have multiple dependsOn tasks.

Reply:
Per Steven's comment, I'd prefer to see unit tests for this kind of difficult-to-understand logic